ratelimit/ratelimit.go

```diff
@@ -0,0 +1,362 @@
+package ratelimit
+
+import (
+	"net"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/NebulousLabs/Sia/build"
+)
+
+// BM is the global bandwidthManager.
+var BM *bandwidthManager
+
+type (
+	// RLConn is a helper struct that wraps a net.Conn and implements the
+	// net.Conn interface.
+	RLConn struct {
+		mu         sync.Mutex
+		writeChan  chan []byte
+		readChan   chan []byte
+		resultChan chan connResult
+		workSignal chan struct{}
+		conn       net.Conn
+	}
+	// connResult contains the return values of a Read or Write.
+	connResult struct {
+		n   int
+		err error
+	}
+	// bandwidthManager is a singleton that coordinates the RLConnections in
+	// the background to guarantee a fair bandwidth distribution over all the
+	// connections.
+	bandwidthManager struct {
+		mu               sync.Mutex
+		conns            []*RLConn
+		atomicWritePPS   int64
+		atomicReadPPS    int64
+		atomicPacketSize int64
+		readSignal       chan struct{}
+		writeSignal      chan struct{}
+		threadRunning    bool
+	}
+)
+
+// Init initializes the bandwidth manager. The first call to Init initializes the bandwidthManager object. Subsequent calls to Init will change the global limits.
+func Init(writePPS, readPPS, packetSize int64, cancel chan struct{}) {
+	// Check if BM already exists
+	if BM != nil {
+		build.Critical("bandwidth manager is already initialized")
+		return
+	}
+	BM = &bandwidthManager{
+		atomicWritePPS:   writePPS,
+		atomicReadPPS:    readPPS,
+		atomicPacketSize: packetSize,
+		writeSignal:      make(chan struct{}),
+		readSignal:       make(chan struct{}),
+	}
+	go BM.threadedWriteLoop(cancel)
+	go BM.threadedReadLoop(cancel)
+}
```

Comment:
if you rename to `init` this function will be run automatically

Reply:
probably makes more sense to do it that way.